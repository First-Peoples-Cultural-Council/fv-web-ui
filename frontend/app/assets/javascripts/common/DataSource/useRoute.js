import { useSelector } from 'react-redux'
function useRoute() {
  return {
    routeParams: useSelector((state) => state.navigation.route.routeParams),
  }
}
export default useRoute
