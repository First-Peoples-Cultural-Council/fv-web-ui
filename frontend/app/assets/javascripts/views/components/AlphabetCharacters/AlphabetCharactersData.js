import { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import ProviderHelpers from 'common/ProviderHelpers'
import selectn from 'selectn'
import { getDialectClassname } from 'views/pages/explore/dialect/helpers'
// REDUX: actions/dispatch/func
import { fetchDocument } from 'providers/redux/reducers/document'
import { fetchCharacters } from 'providers/redux/reducers/fvCharacter'
import { searchDialectUpdate } from 'providers/redux/reducers/searchDialect'
import { pushWindowPath } from 'providers/redux/reducers/windowPath'

import NavigationHelpers from 'common/NavigationHelpers'

class AlphabetCharactersData extends Component {
  constructor(props) {
    super(props)

    this.state = {}
  }

  async componentDidMount() {
    window.addEventListener('popstate', this.clickLetterIfInRouteParams)

    const { routeParams, computePortal } = this.props
    const path = `${routeParams.dialect_path}/Alphabet`
    await ProviderHelpers.fetchIfMissing(
      path,
      this.props.fetchCharacters,
      this.props.computeCharacters,
      '&currentPageIndex=0&pageSize=100&sortOrder=asc&sortBy=fvcharacter:alphabet_order'
    )
    const extractComputedCharacters = ProviderHelpers.getEntry(this.props.computeCharacters, path)
    const characters = selectn('response.entries', extractComputedCharacters)
    const extractComputePortal = ProviderHelpers.getEntry(computePortal, `${routeParams.dialect_path}/Portal`)
    const dialectClassName = getDialectClassname(extractComputePortal)
    this.setState(
      {
        characters,
        dialectClassName,
      },
      () => {
        // this.clickLetterIfInRouteParams()
      }
    )
  }

  // componentDidUpdate(prevProps) {
  //   const letter = selectn('letter', this.props.routeParams)
  //   const prevLetter = selectn('letter', prevProps.routeParams)
  //   if (letter && letter !== prevLetter) {
  //     this.clickLetterIfInRouteParams()
  //   }
  // }

  componentWillUnmount() {
    window.removeEventListener('popstate', this.clickLetterIfInRouteParams)
  }

  render() {
    const { characters, dialectClassName } = this.state
    return this.props.children({
      characters,
      dialectClassName,
      generateAlphabetCharacterHref: this.generateAlphabetCharacterHref,
      activeLetter: this.props.routeParams.letter,
      letterClicked: this.letterClicked,
      splitWindowPath: this.props.splitWindowPath,
    })
  }

  clickLetterIfInRouteParams = () => {
    const letter = selectn('letter', this.props.routeParams)
    if (letter) {
      this.letterClicked({ letter })
    }
  }

  // Used by the presentation layer to generate urls
  generateAlphabetCharacterHref = (letter) => {
    let href = undefined
    const _splitWindowPath = [...this.props.splitWindowPath]
    const wordOrPhraseIndex = _splitWindowPath.findIndex((element) => {
      return element === 'words' || element === 'phrases'
    })
    if (wordOrPhraseIndex !== -1) {
      _splitWindowPath.splice(wordOrPhraseIndex + 1)
      href = `/${_splitWindowPath.join('/')}/alphabet/${letter}`
    }
    return href
  }

  // Called from the presentation layer when a letter is clicked
  letterClicked = async ({ href, letter, updateHistory = false }) => {
    this.props.letterClickedCallback({
      href,
      letter,
      updateHistory,
    })

    if (updateHistory === false && href) {
      NavigationHelpers.navigate(href, this.props.pushWindowPath, false)
    }
  }
}

// PROPTYPES
const { any, array, func, object } = PropTypes
AlphabetCharactersData.propTypes = {
  children: any,
  letterClickedCallback: func,
  // REDUX: reducers/state
  computeCharacters: object.isRequired,
  computeLogin: object.isRequired,
  computePortal: object.isRequired,
  routeParams: object.isRequired,
  splitWindowPath: array.isRequired,
  // REDUX: actions/dispatch/func
  fetchCharacters: func.isRequired,
  fetchDocument: func.isRequired,
  pushWindowPath: func.isRequired,
  searchDialectUpdate: func,
}

AlphabetCharactersData.defaultProps = {
  letterClickedCallback: () => {},
  searchDialectUpdate: () => {},
}

// REDUX: reducers/state
const mapStateToProps = (state) => {
  const { fvCharacter, fvPortal, navigation, nuxeo, windowPath } = state
  const { computePortal } = fvPortal
  const { route } = navigation
  const { computeLogin } = nuxeo
  const { computeCharacters } = fvCharacter
  const { splitWindowPath } = windowPath
  return {
    computePortal,
    computeCharacters,
    computeLogin,
    routeParams: route.routeParams,
    splitWindowPath,
  }
}
// REDUX: actions/dispatch/func
const mapDispatchToProps = {
  fetchDocument,
  fetchCharacters,
  pushWindowPath,
  searchDialectUpdate,
}

export default connect(mapStateToProps, mapDispatchToProps)(AlphabetCharactersData)
