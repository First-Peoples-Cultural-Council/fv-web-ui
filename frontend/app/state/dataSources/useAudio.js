import { useDispatch, useSelector } from 'react-redux'
import { setPlaying as _setPlaying, stopPlaying as _stopPlaying } from 'reducers/audio'
function useAudio() {
  const dispatch = useDispatch()
  return {
    lastPlayed: useSelector((state) => state.audio.lastPlayed),
    setLastPlayed: (src) => {
      dispatch(_setPlaying(src))
    },
    clearLastPlayed: () => {
      dispatch(_stopPlaying())
    },
  }
}
export default useAudio
