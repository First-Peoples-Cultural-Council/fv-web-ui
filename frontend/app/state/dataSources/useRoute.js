import { useDispatch, useSelector } from 'react-redux'
import { setRouteParams as _setRouteParams } from 'reducers/navigation'

function useRoute() {
  const dispatch = useDispatch()

  const setRouteParams = (data) => {
    const dispatchObj = _setRouteParams(data)
    dispatch(dispatchObj)
  }

  return {
    routeParams: useSelector((state) => state.navigation.route.routeParams),
    navigationRouteSearch: useSelector((state) => state.navigation.route.search),
    setRouteParams,
  }
}

export default useRoute
