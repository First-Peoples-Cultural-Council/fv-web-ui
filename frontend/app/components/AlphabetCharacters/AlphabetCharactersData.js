import { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import ProviderHelpers from 'common/ProviderHelpers'
import selectn from 'selectn'
// Data Sources
import useCharacters from 'dataSources/useCharacters'
import usePortal from 'dataSources/usePortal'
import useRoute from 'dataSources/useRoute'
// Helpers
import { getDialectClassname } from 'common/Helpers'
import useNavigationHelpers from 'common/useNavigationHelpers'
/**
 * @summary AlphabetCharactersData
 * @version 2.0.0
 * @component
 *
 * @param {object} props
 *
 * @returns {node} jsx markup
 */
function AlphabetCharactersData({ children }) {
  const [characters, setCharacters] = useState([])
  const { getSearchAsObject, navigate } = useNavigationHelpers()
  const { fetchCharacters, computeCharacters, cacheComputeCharacters } = useCharacters()
  const { computePortal, cacheComputePortal } = usePortal()
  const { routeParams } = useRoute()
  const { letter: queryLetter } = getSearchAsObject()
  const alphabetPath = `${routeParams.dialect_path}/Alphabet`
  const portalPath = `${routeParams.dialect_path}/Portal`

  const extractComputePortal = ProviderHelpers.getEntry(computePortal, portalPath, cacheComputePortal)

  useEffect(() => {
    ProviderHelpers.fetchIfMissing({
      action: fetchCharacters,
      actionParams: '&currentPageIndex=0&pageSize=100&sortOrder=asc&sortBy=fvcharacter:alphabet_order',
      key: alphabetPath,
      reducer: computeCharacters,
      reducerCache: cacheComputePortal,
    })
  }, [])

  const extractComputedCharacters = ProviderHelpers.getEntry(computeCharacters, alphabetPath, cacheComputeCharacters)
  const charactersUnprocessed = selectn('response.entries', extractComputedCharacters) || []
  useEffect(() => {
    const charactersProcessed = charactersUnprocessed.map(({ title }) => {
      return {
        title,
        href: `${window.location.pathname}?letter=${title}`,
        isActiveCharacter: title === decodeURIComponent(queryLetter),
      }
    })
    if (charactersProcessed.length > 0) {
      setCharacters(charactersProcessed)
    }
  }, [charactersUnprocessed, queryLetter])

  const onClick = (href) => {
    navigate(href)
  }

  return children({
    characters,
    dialectClassName: getDialectClassname(extractComputePortal),
    onClick,
  })
}

// PROPTYPES
const { func } = PropTypes
AlphabetCharactersData.propTypes = {
  children: func,
}

export default AlphabetCharactersData
