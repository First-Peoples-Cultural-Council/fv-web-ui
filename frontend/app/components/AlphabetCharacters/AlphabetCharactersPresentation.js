import React from 'react'
import PropTypes from 'prop-types'

import Typography from '@material-ui/core/Typography'
import CircularProgress from '@material-ui/core/CircularProgress'

import FVLabel from 'components/FVLabel'
import '!style-loader!css-loader!./AlphabetCharacters.css'

/**
 * @summary AlphabetCharactersPresentation
 * @version 2.0.0
 * @component
 *
 * @param {object} props
 *
 * @returns {node} jsx markup
 */
function AlphabetCharactersPresentation({
  activeLetter,
  characters = [],
  dialectClassName,
  generateAlphabetCharacterHref,
  letterClicked,
}) {
  const componentHasContent = () => {
    const charactersMarkup = characters.map((value, index) => {
      const currentLetter = value.title
      const href = generateAlphabetCharacterHref(currentLetter)
      return (
        <a
          href={href}
          className={`AlphabetCharactersTile ${activeLetter === currentLetter ? 'AlphabetCharactersTile--active' : ''}`}
          onClick={(e) => {
            e.preventDefault()
            letterClicked({ letter: currentLetter, href })
          }}
          key={index}
        >
          {currentLetter}
        </a>
      )
    })
    return charactersMarkup.length > 0 ? (
      <div className={`AlphabetCharactersTiles ${dialectClassName}`}>{charactersMarkup}</div>
    ) : null
  }

  const componentHasNoContent = () => {
    return (
      <Typography className="AlphabetCharacters__noCharacters" variant="caption">
        Characters are unavailable at this time
      </Typography>
    )
  }

  const componentIsLoading = () => {
    return (
      <div className="AlphabetCharacters__loading">
        <CircularProgress className="AlphabetCharacters__loadingSpinner" color="secondary" mode="indeterminate" />
        <Typography className="AlphabetCharacters__loadingText" variant="caption">
          Loading characters
        </Typography>
      </div>
    )
  }
  return (
    <div className="AlphabetCharacters" data-testid="AlphabetCharacters">
      <h2>
        <FVLabel
          transKey="views.pages.explore.dialect.learn.words.find_by_alphabet"
          defaultStr="Browse Alphabetically"
          transform="words"
        />
      </h2>
      {() => {
        if (characters === undefined) {
          return componentIsLoading()
        }
        if (characters.length === 0) {
          return componentHasNoContent()
        }
        return componentHasContent()
      }}
    </div>
  )
}

// PropTypes
const { array, func, string } = PropTypes
AlphabetCharactersPresentation.propTypes = {
  characters: array,
  dialectClassName: string,
  generateAlphabetCharacterHref: func,
  letterClicked: func,
  activeLetter: string,
}
AlphabetCharactersPresentation.defaultProps = {
  letterClicked: () => {},
  fetchCharacters: () => {},
}

export default AlphabetCharactersPresentation
