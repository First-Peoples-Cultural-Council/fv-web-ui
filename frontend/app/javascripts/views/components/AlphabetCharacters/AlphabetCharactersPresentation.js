import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Typography from '@material-ui/core/Typography'
import CircularProgress from '@material-ui/core/CircularProgress'

import FVLabel from 'views/components/FVLabel'
import '!style-loader!css-loader!./AlphabetCharacters.css'

export class AlphabetCharactersPresentation extends Component {
  render() {
    let content = null
    if (this.props.characters === undefined) {
      content = this.componentIsLoading()
    } else {
      if (this.props.characters.length === 0) {
        content = this.componentHasNoContent()
      } else {
        content = this.componentHasContent()
      }
    }
    return (
      <div className="AlphabetCharacters" data-testid="AlphabetCharacters">
        <h2>
          <FVLabel
            transKey="views.pages.explore.dialect.learn.words.find_by_alphabet"
            defaultStr="Browse Alphabetically"
            transform="words"
          />
        </h2>
        {content}
      </div>
    )
  }

  componentHasContent = () => {
    const { characters = [], generateAlphabetCharacterHref } = this.props
    const { activeLetter } = this.props
    const charactersMarkup = characters.map((value, index) => {
      const currentLetter = value.title
      const href = generateAlphabetCharacterHref(currentLetter)
      return (
        <a
          href={href}
          className={`AlphabetCharactersTile ${activeLetter === currentLetter ? 'AlphabetCharactersTile--active' : ''}`}
          onClick={(e) => {
            e.preventDefault()
            this.props.letterClicked({ letter: currentLetter, href })
          }}
          key={index}
        >
          {currentLetter}
        </a>
      )
    })
    let content = null
    if (charactersMarkup.length > 0) {
      content = <div className={`AlphabetCharactersTiles ${this.props.dialectClassName}`}>{charactersMarkup}</div>
    }
    return content
  }

  componentHasNoContent = () => {
    return (
      <Typography className="AlphabetCharacters__noCharacters" variant="caption">
        Characters are unavailable at this time
      </Typography>
    )
  }

  componentIsLoading = () => {
    return (
      <div className="AlphabetCharacters__loading">
        <CircularProgress className="AlphabetCharacters__loadingSpinner" color="secondary" mode="indeterminate" />
        <Typography className="AlphabetCharacters__loadingText" variant="caption">
          Loading characters
        </Typography>
      </div>
    )
  }
}

// PropTypes
const { array, func, string } = PropTypes
AlphabetCharactersPresentation.propTypes = {
  characters: array,
  dialectClassName: string,
  generateAlphabetCharacterHref: func,
  letterClicked: func,
  activeLetter: string,
}
AlphabetCharactersPresentation.defaultProps = {
  letterClicked: () => {},
  fetchCharacters: () => {},
}

export default AlphabetCharactersPresentation
