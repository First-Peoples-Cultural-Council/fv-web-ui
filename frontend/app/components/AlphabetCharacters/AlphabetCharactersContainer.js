import React from 'react'
// import PropTypes from 'prop-types'
import AlphabetCharactersPresentation from 'components/AlphabetCharacters/AlphabetCharactersPresentation'
import AlphabetCharactersData from 'components/AlphabetCharacters/AlphabetCharactersData'

/**
 * @summary AlphabetCharactersContainer
 * @version 2.0.0
 * @component
 *
 * @param {object} props
 *
 * @returns {node} jsx markup
 */
function AlphabetCharactersContainer() {
  return (
    <AlphabetCharactersData>
      {({ characters, dialectClassName, onClick }) => {
        return (
          <AlphabetCharactersPresentation
            characters={characters}
            dialectClassName={dialectClassName}
            onClick={onClick}
          />
        )
      }}
    </AlphabetCharactersData>
  )
}
// PROPTYPES
// const { string } = PropTypes
AlphabetCharactersContainer.propTypes = {
  //   something: string,
}

export default AlphabetCharactersContainer
