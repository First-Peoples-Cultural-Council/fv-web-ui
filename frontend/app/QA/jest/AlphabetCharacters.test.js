// Mock: react-redux
// --------------------------------------------------
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}))

// Mock: @material-ui/core/styles
// --------------------------------------------------
jest.mock('@material-ui/core/styles', () => ({
  /* eslint-disable-next-line */
  withStyles: (styles) => (component) => component,
  withTheme: () => {},
}))

// Mock: useRoute
// --------------------------------------------------
import * as useRoute from 'dataSources/useRoute'
const spyUseRoute = jest.spyOn(useRoute, 'default')
spyUseRoute.mockReturnValue({
  routeParams: {
    letter: '',
    category: '',
    phraseBook: '',
  },
})

// Let us begin!
// ==================================================
import React from 'react'
import { Provider } from 'react-redux'
import store from 'state/store'
import renderer from 'react-test-renderer'
import AlphabetCharactersContainer from 'components/AlphabetCharacters/AlphabetCharactersContainer'
import AlphabetCharactersData from 'components/AlphabetCharacters/AlphabetCharactersData'
import AlphabetCharactersPresentation from 'components/AlphabetCharacters/AlphabetCharactersPresentation'

describe('AlphabetCharacters', () => {
  // Container
  test('Container', () => {
    const component = renderer.create(
      <Provider store={store}>
        <AlphabetCharactersContainer />
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  // Data
  test('Data', () => {
    // Updated in AlphabetCharactersData.children
    let output = {}

    renderer.create(
      <AlphabetCharactersData>
        {(dataLayerOutput) => {
          output = dataLayerOutput
          return null
        }}
      </AlphabetCharactersData>
    )
    expect(output).toMatchInlineSnapshot(`
      Object {
        "characters": Array [],
        "dialectClassName": "fontBCSans",
        "onClick": [Function],
      }
    `)
  })
  // Presentation
  test('Presentation > 1 Active & 1 Inactive', () => {
    const component = renderer.create(
      <Provider store={store}>
        <AlphabetCharactersPresentation
          characters={[
            { title: 'Active', href: '#', isActiveCharacter: true },
            { title: 'Inctive', href: '#', isActiveCharacter: false },
          ]}
        />
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  // Presentation
  test('Presentation > No Chars', () => {
    const component = renderer.create(
      <Provider store={store}>
        <AlphabetCharactersPresentation characters={[]} />
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
