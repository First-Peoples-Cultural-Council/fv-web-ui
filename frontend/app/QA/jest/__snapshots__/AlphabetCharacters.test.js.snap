// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlphabetCharacters Container 1`] = `
<div
  className="AlphabetCharacters"
  data-testid="AlphabetCharacters"
>
  <h2>
    <span
      className="fv-label"
    >
      Browse Alphabetically
    </span>
  </h2>
  <span
    className="MuiTypography-root AlphabetCharacters__noCharacters MuiTypography-caption"
  >
    Characters are unavailable at this time
  </span>
</div>
`;

exports[`AlphabetCharacters Presentation > 1 Active & 1 Inactive 1`] = `
<div
  className="AlphabetCharacters"
  data-testid="AlphabetCharacters"
>
  <h2>
    <span
      className="fv-label"
    >
      Browse Alphabetically
    </span>
  </h2>
  <div
    className="AlphabetCharactersTiles "
  >
    <a
      className="AlphabetCharactersTile AlphabetCharactersTile--active"
      href="#"
      onClick={[Function]}
    >
      Active
    </a>
    <a
      className="AlphabetCharactersTile "
      href="#"
      onClick={[Function]}
    >
      Inctive
    </a>
  </div>
</div>
`;

exports[`AlphabetCharacters Presentation > No Chars 1`] = `
<div
  className="AlphabetCharacters"
  data-testid="AlphabetCharacters"
>
  <h2>
    <span
      className="fv-label"
    >
      Browse Alphabetically
    </span>
  </h2>
  <span
    className="MuiTypography-root AlphabetCharacters__noCharacters MuiTypography-caption"
  >
    Characters are unavailable at this time
  </span>
</div>
`;
